### ISSUE TITLE
After succesfull home the machine first move is to Z=0, this triggers alarm

### ISSUE THREAD

**Issue by i-g-g-y** [grbl/grbl#1776]
Hello.

I work with F360 and very first move in code is:

`G90 G94
G17
G21
(When using Fusion 360 for Personal Use, the feedrate of)
(rapid moves is reduced to match the feedrate of cutting)
(moves, which can increase machining time. Unrestricted rapid)
(moves are available with a Fusion 360 Subscription.)
G28 G91 Z0
G90

(Face12)
T1
S1000 M3
`

This of course triggers my Z axis endpoint switch and raises an alarm.

How can I put my Z switch at Z=1mm, so a move to Z=0 does not trigger the switch?

**Comment by langwadt** [comment]
define HOMING_FORCE_SET_ORIGIN (line 87 in config.h) it'll set the machine position to 0,0,0 after the switch pull off

alternatively just remove the lines with G28 (you can also turn them off in the post processor)

**Comment by i-g-g-y** [comment]
#define HOMING_FORCE_SET_ORIGIN

did not work.

**Comment by langwadt** [comment]
are you sure you changed, build, and flashed the right code?

what is the machine position after homing?

**Comment by i-g-g-y** [comment]
Yes, I am sure, as I used the same process to disable the spindle control, so I can have all 3 switches.

**Comment by langwadt** [comment]
what is the machine position after homing?

**Comment by i-g-g-y** [comment]
Man I just damaged my machine. I have to change one of my switches. I have to disassemble the complete machine.

**Comment by i-g-g-y** [comment]
I fixed my machine.

Here is the screen after homing:
![image](https://user-images.githubusercontent.com/10361323/156940652-54feff99-eb81-4c68-8a9d-ba59ab394f56.png)

And the soft limits:
![image](https://user-images.githubusercontent.com/10361323/156940668-d5194579-811b-42c5-a1ab-1c2771944af9.png)


Config for homing:
![image](https://user-images.githubusercontent.com/10361323/156940684-5b1d9d31-96af-47f4-973a-270a6a9df398.png)

**Comment by i-g-g-y** [comment]
>>> $$
$0 = 10    (Step pulse time, microseconds)
$1 = 25    (Step idle delay, milliseconds)
$2 = 0    (Step pulse invert, mask)
$3 = 4    (Step direction invert, mask)
$4 = 0    (Invert step enable pin, boolean)
$5 = 1    (Invert limit pins, boolean)
$6 = 0    (Invert probe pin, boolean)
$10 = 16    (Status report options, mask)
$11 = 0.010    (Junction deviation, millimeters)
$12 = 0.002    (Arc tolerance, millimeters)
$13 = 0    (Report in inches, boolean)
$20 = 1    (Soft limits enable, boolean)
$21 = 1    (Hard limits enable, boolean)
$22 = 1    (Homing cycle enable, boolean)
$23 = 3    (Homing direction invert, mask)
$24 = 100.000    (Homing locate feed rate, mm/min)
$25 = 500.000    (Homing search seek rate, mm/min)
$26 = 200    (Homing switch debounce delay, milliseconds)
$27 = 5.000    (Homing switch pull-off distance, millimeters)
$30 = 1000    (Maximum spindle speed, RPM)
$31 = 0    (Minimum spindle speed, RPM)
$32 = 0    (Laser-mode enable, boolean)
$100 = 400.000    (X-axis travel resolution, step/mm)
$101 = 400.000    (Y-axis travel resolution, step/mm)
$102 = 161.000    (Z-axis travel resolution, step/mm)
$110 = 1000.000    (X-axis maximum rate, mm/min)
$111 = 1000.000    (Y-axis maximum rate, mm/min)
$112 = 500.000    (Z-axis maximum rate, mm/min)
$120 = 10.000    (X-axis acceleration, mm/sec^2)
$121 = 10.000    (Y-axis acceleration, mm/sec^2)
$122 = 10.000    (Z-axis acceleration, mm/sec^2)
$130 = 277.000    (X-axis maximum travel, millimeters)
$131 = 190.000    (Y-axis maximum travel, millimeters)
$132 = 160.000    (Z-axis maximum travel, millimeters)
ok

**Comment by langwadt** [comment]
ah you are homing in the reverse direction, then I don't know if HOMING_FORCE_SET_ORIGIN will work then, never tried. 

you could change, line 302-304 in limits.c, removing the pull off

**Comment by i-g-g-y** [comment]
Don't I need this pull-off?

**Comment by langwadt** [comment]
the pull off is part of the homing cycle, but those lines set the position after the homing is all done. i.e. with you 5mm pulloff on Z it sets the Z position to -5mm after homing is done, what you want and imo what makes most sense it to set Z = 0mm after homing is done

**Comment by i-g-g-y** [comment]
OK, can you give me the exact syntax to change? I have GRBL 1.1h and my line numbers do not correspond to the ones you are giving me.

**Comment by i-g-g-y** [comment]
![image](https://user-images.githubusercontent.com/10361323/156942481-a1bdc411-67aa-4d90-9be8-d0eadfc00f20.png)

**Comment by langwadt** [comment]
ok, then you should really be in https://github.com/gnea/grbl/ not this one

anyway:    remove the HOMING_FORCE_SET_ORIGIN define, that won't work with reverse homing 
change line 382 in limits.c from:
`set_axis_position = lround(-settings.homing_pulloff*settings.steps_per_mm[idx]);`
to:
`set_axis_position = 0`

then you Z position will be 0mm after moving away from the switch, instead of -5mm

**Comment by i-g-g-y** [comment]
Still the same??

![image](https://user-images.githubusercontent.com/10361323/156943460-91847f93-9ca6-4767-8b2d-0cb6a45d5b99.png)

![image](https://user-images.githubusercontent.com/10361323/156943483-67829b34-bef4-4997-bbd1-e8567c0d661e.png)

**Comment by langwadt** [comment]
is it Z or X/Y that gets triggered?

what G28 Z0 does is move to Z=0 and the to the position stored for G28 if you haven't stored a position that is 0,0,0
 
move the spindle to a convenient safe place and store that position with G28.1

**Comment by i-g-g-y** [comment]
It is Z axis.

**Comment by langwadt** [comment]
what is the Z position after homing is done? it should 0 and not touching the switch

**Comment by i-g-g-y** [comment]
Nope. How do I figure out where is my code coming from? Looks like I am modifying code that is not used? I am completely confused now.

![image](https://user-images.githubusercontent.com/10361323/156944001-c7893b32-7db7-4087-b793-969e41003f85.png)

**Comment by i-g-g-y** [comment]
> ok, then you should really be in https://github.com/gnea/grbl/ not this one

Sorry, what?

**Comment by i-g-g-y** [comment]
> ok, then you should really be in https://github.com/gnea/grbl/ not this one

I really do not understand what you ment?

**Comment by langwadt** [comment]
https://github.com/grbl/grbl/ is the repo for older versions of grbl, newer version like the one you use is on https://github.com/gnea/grbl/  https://github.com/grbl/grbl/wiki#grbl-v11-has-been-released-at-our-new-site-the-old-site-will-eventually-be-phased-out

I haven't build it in arduino, but I've heard of people editing the wrong files

https://github.com/grbl/grbl/wiki/Compiling-Grbl

under the heading "Advanced Users: " might help
 
when not sure if you are editing the right file,  add some random characters and verify that you get a compile error

**Comment by i-g-g-y** [comment]
I have found the right code to edit. So I did. Editing the code completely screws up the machine. It would unscrew the Y axis if I let it, so have to do an emergency stop. I am going back to original code.

No dice. As soon I tell the machine to go to work zero it throws an error: 

![image](https://user-images.githubusercontent.com/10361323/156945378-1642365c-4b29-4fec-9a24-d9cb01cdc912.png)


This is how it looks like after homing:

![image](https://user-images.githubusercontent.com/10361323/156945419-8e9809e7-d4bb-448b-a90f-8b1a6cefc16c.png)

**Comment by langwadt** [comment]
work position can be anywhere you set it to, homing set machine coordinates, work coordinates are set realtive to that when you zeros at some postition where you want work zero 

the error is not that you hit the switch(hard limit), it is that you moved outside the machines defined limits (soft limit)

your $132 = 160.000 means you Z axis can be in the range -160mm to 0mm, you are at 0 in machines coordinates and 2.7mm work coordinates , G0Z5 tells it to move to 5mm in work coordinates, which would be 2.3mm in machine coordinates, that is clearly outside the 0 to -160mm range

you should move the spindle to where your work zero should be and zero all axis first

**Comment by i-g-g-y** [comment]
That is the thing. If I modify the code, it wants to unscrew itself and have to do an emergency stop. I do set work coordinates. I will show tomorrow.

**Comment by i-g-g-y** [comment]
Hello.

With this code modification it works, I do not know what I was doing wrong. As a feature, I would request the limit switches be put out of bound of the moving area of the machine.

![image](https://user-images.githubusercontent.com/10361323/157122980-697aa5dc-501b-4fc3-9ab7-b381b7dc81a1.png)

**Comment by i-g-g-y** [comment]
Here a picture, so you can see where does the progress go. I have to modify the Y bracket, tidy this up, but it does work and it chews steel, but only little bites.

![image](https://user-images.githubusercontent.com/10361323/157124272-fcb89c27-14b0-4699-ae43-9402ef84f9da.png)
